polkadot/xcm/pallet-xcm/src/lib.rs

```diff
@@ -1194,13 +1212,42 @@ impl<T: Config> QueryHandler for Pallet<T> {
 }
 
 impl<T: Config> Pallet<T> {
+	/// Validate `assets` to be reserve-transferred and return their reserve location.
+	fn validate_assets_and_find_reserve(
+		assets: &[MultiAsset],
+		dest: &MultiLocation,
+	) -> Result<TransferType, Error<T>> {
+		let mut reserve = None;
+		for asset in assets.iter() {
+			// Ensure fungible asset.
+			ensure!(
```

Comment:
I am not sure about this `Funbiglity` check, iiuc previous version of `do_reserve_transfer_assets` did not check that and also `XcmExecutor` does not check fungibility for `TransferReserveAsset` - it just depends on provided `AssetTransactor`. 

I would probably just check `is_zero` only for `Fungible` (make sense not to allow zero transfer) and I would allow to pass `Fungibility::NonFungible`. Runtime can filter-out non-fungibles with `T::XcmReserveTransferFilter` or `AssetTransactor`.

Reply:
fixed